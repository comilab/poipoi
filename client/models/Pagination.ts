export default class Pagination {
  currentPage: number
  from: number
  lastPage: number
  path: string
  perPage: number
  to: number
  total: number

  constructor (data: any) {
    this.currentPage = data.currentPage
    this.from = data.from
    this.lastPage = data.lastPage
    this.path = data.path
    this.perPage = data.perPage
    this.to = data.to
    this.total = data.total
  }
}
