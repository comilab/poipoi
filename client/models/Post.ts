import dayjs, { Dayjs } from 'dayjs'

import Image from '~/models/Image'
import Reaction from '~/models/Reaction'

export type ScopeUnion = 'public'|'password'|'private'

export type RatingUnion = null|'nsfw'|'r18'

export default class Post {
  id: number
  userId: number
  title: string
  caption: string
  showImagesList: boolean
  text: string
  scope: ScopeUnion
  password: string|null
  publishStart: Dayjs|null = null
  publishEnd: Dayjs|null = null
  rating: RatingUnion
  pinned: boolean
  showThumbnail: boolean
  denyRobot: boolean|null
  enableReaction: boolean|null
  allowedEmojis: string[]|null
  deniedEmojis: string[]|null
  enableTwitterShare: boolean|null
  createdAt: Dayjs|null = null
  updatedAt: Dayjs|null = null

  url: string
  textCount: number
  actualDenyRobot: boolean
  actualEnableReaction: boolean
  actualAllowedEmojis: string[]
  actualDeniedEmojis: string[]
  actualEnableTwitterShare: boolean
  emojis: string[] = []

  images: Image[] = []
  imagesCount: number

  reactions: Reaction[] = []
  reactionsCount: number

  constructor (data: any) {
    this.id = data.id
    this.userId = data.userId
    this.title = data.title
    this.caption = data.caption
    this.showImagesList = data.showImagesList
    this.text = data.text
    this.scope = data.scope
    this.password = data.password
    if (data.publishStart) {
      this.publishStart = dayjs.utc(data.publishStart).local()
    }
    if (data.publishEnd) {
      this.publishEnd = dayjs.utc(data.publishEnd).local()
    }
    this.rating = data.rating
    this.pinned = data.pinned
    this.showThumbnail = data.showThumbnail
    this.denyRobot = data.denyRobot
    this.enableReaction = data.enableReaction
    this.allowedEmojis = data.allowedEmojis
    this.deniedEmojis = data.deniedEmojis
    this.enableTwitterShare = data.enableTwitterShare
    if (data.createdAt) {
      this.createdAt = dayjs.utc(data.createdAt).local()
    }
    if (data.updatedAt) {
      this.updatedAt = dayjs.utc(data.updatedAt).local()
    }

    this.url = data.url
    this.textCount = data.textCount
    this.actualDenyRobot = data.actualDenyRobot
    this.actualEnableReaction = data.actualEnableReaction
    this.actualAllowedEmojis = data.actualAllowedEmojis
    this.actualDeniedEmojis = data.actualDeniedEmojis
    this.actualEnableTwitterShare = data.actualEnableTwitterShare
    this.emojis = data.emojis

    if (data.images) {
      this.images = data.images.map((image: any) => new Image(image))
    }
    this.imagesCount = data.imagesCount

    if (data.reactions) {
      this.reactions = data.reactions.map((reaction: any) => new Reaction(reaction))
    }
    this.reactionsCount = data.reactionsCount
  }

  get path () {
    return `/posts/${this.id}`
  }

  get textPages () {
    return (this.text || '').split('[newpage]').map(page => page.trim())
  }
}
