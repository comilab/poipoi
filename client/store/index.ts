import { defineComponent, provide, inject } from 'nuxt-composition-api'
import { InjectionKey } from '@vue/composition-api'

import sessionStore from '~/store/session'
import settingStore from '~/store/setting'
import toastStore from '~/store/toast'

type Store = {
  session: ReturnType<typeof sessionStore>
  setting: ReturnType<typeof settingStore>
  toast: ReturnType<typeof toastStore>
}

const makeStore = () => {
  return {
    session: sessionStore(),
    setting: settingStore(),
    toast: toastStore()
  }
}

export const key: InjectionKey<Store> = Symbol('Store')

export const store: Store = makeStore()

export const StoreProvider = defineComponent({
  setup () {
    provide(key, store)
  },
  render (h) {
    return h('div', this.$slots.default)
  }
})

export const useStore = () => inject(key)!

export const resetStore = () => {
  Object.assign(store, makeStore())
}

export const actions = {
  async nuxtClientInit () {
    try {
      await Promise.all([
        store.session.load(),
        store.setting.load()
      ])
    } catch (error) {
      //
    }
  }
}
