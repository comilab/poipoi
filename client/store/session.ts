import { reactive, toRefs, computed } from 'nuxt-composition-api'
import axios from 'axios'

import User from '~/models/User'

export default function sessionStore () {
  const state = reactive({
    user: null as User|null
  })

  const isLogin = computed(() => !!state.user)

  async function create (query: any) {
    await axios.get('/sanctum/csrf-cookie', {
      baseURL: process.env.apiUrl
    })
    await axios.post('/sessions', {
      ...query,
      device_name: 'poipoi-web'
    })
  }

  async function load () {
    try {
      const { data } = await axios.get('/sessions')
      setUser(new User(data))
    } catch (error) {
      setUser(null)
    }
  }

  async function destroy () {
    try {
      await axios.delete('/sessions/poipoi-web')
    } catch (error) {
      //
    }
    setUser(null)
  }

  function setUser (user: User|null) {
    state.user = user
  }

  return {
    ...toRefs(state),
    isLogin,
    create,
    load,
    destroy,
    setUser
  }
}
