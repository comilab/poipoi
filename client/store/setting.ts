import { reactive, toRefs } from 'nuxt-composition-api'
import axios from 'axios'

import useApi from '~/composables/use-api'
import Setting from '~/models/Setting'

export default function sessionStore () {
  const api = useApi('/settings', Setting)

  const state = reactive({
    data: null as Setting|null
  })

  async function load () {
    const { data } = await axios.get('/settings')
    setData(new Setting(data))
  }

  async function save (data: any) {
    await api.save(data)
    setData(api.item.value!)
  }

  function setData (data: Setting|null) {
    state.data = data
  }

  return {
    ...toRefs(state),
    load,
    save,
    setData
  }
}
