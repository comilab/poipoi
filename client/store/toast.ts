import { reactive, toRefs } from 'nuxt-composition-api'

export type Toast = {
  type: 'success'|'info'|'warning'|'danger'
  message: string
  timeout?: number
}

export default function toastStore () {
  const state = reactive({
    toasts: [] as Toast[]
  })

  function push ({ type, message, timeout = 5000 }: Toast) {
    const toast = {
      type,
      message,
      timeout
    }
    state.toasts.push(toast)
    if (timeout) {
      setTimeout(() => {
        remove(toast)
      }, timeout)
    }
  }

  function remove (toast: Toast) {
    const index = state.toasts.findIndex(t => t === toast)
    if (index >= 0) {
      state.toasts.splice(index, 1)
    }
  }

  return {
    ...toRefs(state),
    push,
    remove
  }
}
