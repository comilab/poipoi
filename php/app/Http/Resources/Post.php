<?php

namespace App\Http\Resources;

use Illuminate\Http\Resources\Json\JsonResource;

class Post extends JsonResource
{
    /**
     * Transform the resource into an array.
     *
     * @param  \Illuminate\Http\Request  $request
     * @return array
     */
    public function toArray($request)
    {
        return [
            'id' => $this->id,
            'userId' => $this->user_id,
            'title' => $this->title,
            'caption' => $this->caption,
            'showImagesList' => $this->show_images_list,
            'text' => $this->text,
            'scope' => $this->scope,
            'password' => $this->when(auth()->check(), $this->password),
            'publishStart' => $this->publish_start,
            'publishEnd' => $this->publish_end,
            'rating' => $this->rating,
            'pinned' => $this->pinned,
            'showThumbnail' => $this->show_thumbnail,
            'denyRobot' => $this->deny_robot,
            'enableReaction' => $this->enable_reaction,
            'allowedEmojis' => $this->allowed_emojis,
            'deniedEmojis' => $this->denied_emojis,
            'enableTwitterShare' => $this->enable_twitter_share,
            'createdAt' => $this->created_at,
            'updatedAt' => $this->updated_at,

            'url' => $this->url,
            'textCount' => $this->text_count,
            'actualDenyRobot' => $this->actual_deny_robot,
            'actualEnableReaction' => $this->actual_enable_reaction,
            'actualAllowedEmojis' => $this->actual_allowed_emojis,
            'actualDeniedEmojis' => $this->actual_denied_emojis,
            'actualEnableTwitterShare' => $this->actual_enable_twitter_share,
            'emojis' => $this->emojis,

            'images' => Image::collection($this->whenLoaded('images')),
            'imagesCount' => (int) $this->images_count,

            'reactions' => Reaction::collection($this->when(
                $this->is_mine || $this->actual_enable_reaction,
                $this->whenLoaded('reactions')
            )),
            'reactionsCount' => $this->when(
                $this->is_mine || $this->actual_enable_reaction,
                (int) $this->reactions_count,
            ),
        ];
    }
}
