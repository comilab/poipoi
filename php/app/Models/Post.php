<?php

namespace App\Models;

use App\Services\Emoji;
use App\Services\Setting as SettingService;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Support\Carbon;
use Illuminate\Support\Collection;
use Illuminate\Support\Facades\Storage;

class Post extends Model
{
    use HasFactory;

    protected $fillable = [
        'title',
        'caption',
        'show_images_list',
        'text',
        'scope',
        'password',
        'publish_start',
        'publish_end',
        'rating',
        'pinned',
        'show_thumbnail',
        'deny_robot',
        'enable_reaction',
        'allowed_emojis',
        'denied_emojis',
        'enable_twitter_share',
    ];

    protected $casts = [
        'user_id' => 'integer',
        'show_images_list' => 'boolean',
        'publish_start' => 'datetime',
        'publish_end' => 'datetime',
        'pinned' => 'boolean',
        'show_thumbnail' => 'boolean',
        'deny_robot' => 'boolean',
        'enable_reaction' => 'boolean',
        'allowed_emojis' => 'collection',
        'denied_emojis' => 'collection',
        'enable_twitter_share' => 'boolean',
    ];

    public function user()
    {
        return $this->belongsTo(User::class);
    }

    public function images()
    {
        return $this->hasMany(Image::class)->orderBy('weight');
    }

    public function reactions()
    {
        return $this->hasMany(Reaction::class)->latest();
    }

    public function tags()
    {
        return $this->belongsToMany(Tag::class);
    }

    public function getUrlAttribute(): string
    {
        return url("/posts/{$this->id}");
    }

    public function getTextCountAttribute(): int
    {
        return mb_strlen($this->text);
    }

    public function getActualDenyRobotAttribute(): bool
    {
        if ($this->deny_robot !== null) {
            return $this->deny_robot;
        }

        $denyRobotScope = app(SettingService::class)->get('postDefault.denyRobotScope');

        return in_array($this->scope, $denyRobotScope, true) || in_array($this->rating, $denyRobotScope, true);
    }

    public function getActualEnableReactionAttribute(): bool
    {
        if ($this->enable_reaction !== null) {
            return $this->enable_reaction;
        }

        return app(SettingService::class)->get('postDefault.enableReaction');
    }

    public function getActualAllowedEmojisAttribute(): Collection
    {
        if ($this->allowed_emojis !== null) {
            return $this->allowed_emojis;
        }

        return collect(app(SettingService::class)->get('postDefault.allowedEmojis'));
    }

    public function getActualDeniedEmojisAttribute(): Collection
    {
        if ($this->denied_emojis !== null) {
            return $this->denied_emojis;
        }

        return collect(app(SettingService::class)->get('postDefault.deniedEmojis'));
    }

    public function getEmojisAttribute(): Collection
    {
        if ($this->actual_allowed_emojis->isNotEmpty() && $this->actual_denied_emojis->isNotEmpty()) {
            return $this->actual_allowed_emojis->reject(function ($emoji) {
                return $this->actual_denied_emojis->contains($emoji);
            });
        } else if ($this->actual_allowed_emojis->isNotEmpty()) {
            return $this->actual_allowed_emojis;
        } else if ($this->actual_denied_emojis->isNotEmpty()) {
            return app(Emoji::class)->all()->reject(function ($emoji) {
                return $this->actual_denied_emojis->contains($emoji);
            });
        } else {
            return collect([]);
        }
    }

    public function getActualEnableTwitterShareAttribute(): bool
    {
        if ($this->enable_twitter_share !== null) {
            return $this->enable_twitter_share;
        }

        return app(SettingService::class)->get('postDefault.enableTwitterShare');
    }

    public function getIsMineAttribute(): bool
    {
        return auth()->check() && (auth()->user()->id === $this->user_id);
    }

    public function getIsPublicAttribute(): bool
    {
        if ($this->scope === 'private') {
            return false;
        }

        if ($this->publish_start && Carbon::now()->isBefore($this->publish_start)) {
            return false;
        }
        if ($this->publish_end && Carbon::now()->isAfter($this->publish_end)) {
            return false;
        }

        return true;
    }

    public function getImagesDirAttribute(): string
    {
        return "public/images/{$this->id}";
    }

    public function scopePublic($query)
    {
        return $query
            ->where('scope', '!=', 'private')
            ->where(function ($query) {
                $query
                    ->orWhere('publish_start', null)
                    ->orWhere('publish_start', '<=', Carbon::now());
            })
            ->where(function ($query) {
                $query
                    ->orWhere('publish_end', null)
                    ->orWhere('publish_end', '>=', Carbon::now());
            });
    }

    public function scopeSearch(Builder $query, $keyword)
    {
        $keywords = preg_split('/[\s]/', $keyword);
        $query->where(function ($query) use ($keywords) {
            foreach ($keywords as $keyword) {
                if ($keyword[0] === '#') {
                    $tag = substr($keyword, 1);
                    $query->whereHas('tags', function (Builder $query) use ($tag) {
                        $query->where('name', $tag);
                    });
                } else {
                    $escaped = str_replace(['\\', '%', '_'], ['\\\\', '\%', '\_'], $keyword);
                    $query->whereRaw("caption like ? escape '\'", ["%{$escaped}%"]);
                }
            }
        });
    }

    public function syncTagsFromCaption()
    {
        preg_match_all('/#(\S+)/', $this->caption, $matches);

        $tags = collect($matches[1])
            ->map(function ($tagName) {
                return Tag::firstOrCreate(['name' => $tagName]);
            });

        $this->tags()->sync($tags->pluck('id'));

        resolve(Tag::class)->clean();

        return $this;
    }

    public function makeImagesDir()
    {
        Storage::makeDirectory($this->images_dir);
    }

    public function deleteImagesDir()
    {
        Storage::deleteDirectory($this->images_dir);
    }
}
