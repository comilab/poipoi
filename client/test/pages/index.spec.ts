import { render, waitFor } from '@testing-library/vue'
import { mocked } from 'ts-jest/utils'
import { ref, reactive } from '@vue/composition-api'
import * as NuxtCompositionApi from 'nuxt-composition-api'
import VueScrollTo from 'vue-scrollto'

import Index from '~/pages/index.vue'
import * as store from '~/store'
import usePostsApi from '~/composables/use-posts-api'
import postFactory from '~/test/factories/post'
import settingFactory from '~/test/factories/setting'
import Post from '~/models/Post'
import Pagination from '~/models/Pagination'
import paginationFactory from '~/test/factories/pagination'
import useMockedComponent from '~/test/stubs/MockedComponent'

jest.mock('vue-scrollto')
jest.mock('~/composables/use-posts-api')

describe('pages/index', () => {
  let options: any
  let contextMock: any
  let metaMock: any
  let fetchMock: any
  let postsApiMock: any

  beforeEach(() => {
    options = {
      props: {
        post: postFactory.build()
      },
      stubs: {
        FontAwesomeIcon: true,
        NuxtLink: true
      }
    }

    store.store.setting.setData(settingFactory.build())
    jest.spyOn(store, 'useStore').mockReturnValue(store.store)

    contextMock = {
      redirect: jest.fn(),
      route: ref<any>({
        query: {}
      })
    }
    jest.spyOn(NuxtCompositionApi, 'useContext').mockReturnValue(contextMock)

    metaMock = {
      title: ref('')
    }
    jest.spyOn(NuxtCompositionApi, 'useMeta').mockReturnValue(metaMock)

    fetchMock = {
      fetch: jest.fn(),
      fetchState: reactive({
        pending: false
      })
    }
    jest.spyOn(NuxtCompositionApi, 'useFetch').mockReturnValue(fetchMock)

    postsApiMock = {
      index: jest.fn(),
      posts: ref<Post[]>([]),
      pagination: ref<Pagination|null>(null)
    }
    mocked(usePostsApi).mockReturnValue(postsApiMock)
  })

  describe('renders correctly', () => {
    test('投稿がある時', () => {
      postsApiMock.posts.value = postFactory.buildList(10, {
        url: 'https://example.com'
      })
      postsApiMock.pagination.value = paginationFactory.build({
        currentPage: 1,
        perPage: 10,
        total: 100
      })
      options.stubs.AppPostThumbnail = true
      options.stubs.AppPagination = true
      const { container } = render(Index, options)

      expect(container).toMatchSnapshot()
    })

    test('投稿がない時', () => {
      postsApiMock.posts.value = []
      postsApiMock.pagination.value = paginationFactory.build({
        currentPage: 1,
        perPage: 10,
        total: 100
      })
      options.stubs.AppPagination = true
      const { container } = render(Index, options)

      expect(container).toMatchSnapshot()
    })
  })

  describe('title', () => {
    test('siteTitleをセット', () => {
      store.store.setting.setData(settingFactory.build({
        siteTitle: 'site-title'
      }))
      render(Index, options)

      expect(metaMock.title.value).toBe('site-title')
    })
  })

  describe('fetch', () => {
    test('indexが実行される', async () => {
      render(Index, options)

      expect(NuxtCompositionApi.useFetch).toBeCalled()

      const callback = mocked(NuxtCompositionApi.useFetch).mock.calls[0][0] as Function
      await callback()

      expect(postsApiMock.index).toBeCalled()
    })

    test('上部にスクロール', async () => {
      render(Index, options)

      expect(NuxtCompositionApi.useFetch).toBeCalled()

      const callback = mocked(NuxtCompositionApi.useFetch).mock.calls[0][0] as Function
      await callback()

      expect(VueScrollTo.scrollTo).toBeCalledWith('#top')
    })

    test('route.queryが変更されたら再実行', async () => {
      render(Index, options)

      contextMock.route.value = {
        query: {
          keyword: 'keyword'
        }
      }

      await waitFor(() => {
        expect(fetchMock.fetch).toBeCalled()
      })
    })
  })

  describe('pagination', () => {
    test('selectedされたらリダイレクトする', async () => {
      postsApiMock.posts.value = postFactory.buildList(1)
      const { component, emitsList } = useMockedComponent({
        emits: ['selected']
      })
      postsApiMock.pagination.value = paginationFactory.build()
      options.stubs.AppPagination = component

      render(Index, options)
      emitsList[0].selected(1)

      await waitFor(() => {
        expect(contextMock.redirect).toBeCalledWith('/', {
          page: '1'
        })
      })
    })
  })
})
