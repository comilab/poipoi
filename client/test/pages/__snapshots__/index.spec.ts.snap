// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/index renders correctly 投稿がある時 1`] = `
<div>
  <div
    post="[object Object]"
  >
    <section
      class="rounded bg-white my-8 p-4"
      data-testid="section"
    >
      <!---->
      <div
        class="grid gap-4 grid-cols-2 md_grid-cols-4 lg_grid-cols-6"
      >
        <nuxtlink-stub
          class="post-link inline-block relative"
          to="/posts/2"
        >
          <apppostthumbnail-stub
            post="[object Object]"
          />
        </nuxtlink-stub>
        <nuxtlink-stub
          class="post-link inline-block relative"
          to="/posts/3"
        >
          <apppostthumbnail-stub
            post="[object Object]"
          />
        </nuxtlink-stub>
        <nuxtlink-stub
          class="post-link inline-block relative"
          to="/posts/4"
        >
          <apppostthumbnail-stub
            post="[object Object]"
          />
        </nuxtlink-stub>
        <nuxtlink-stub
          class="post-link inline-block relative"
          to="/posts/5"
        >
          <apppostthumbnail-stub
            post="[object Object]"
          />
        </nuxtlink-stub>
        <nuxtlink-stub
          class="post-link inline-block relative"
          to="/posts/6"
        >
          <apppostthumbnail-stub
            post="[object Object]"
          />
        </nuxtlink-stub>
        <nuxtlink-stub
          class="post-link inline-block relative"
          to="/posts/7"
        >
          <apppostthumbnail-stub
            post="[object Object]"
          />
        </nuxtlink-stub>
        <nuxtlink-stub
          class="post-link inline-block relative"
          to="/posts/8"
        >
          <apppostthumbnail-stub
            post="[object Object]"
          />
        </nuxtlink-stub>
        <nuxtlink-stub
          class="post-link inline-block relative"
          to="/posts/9"
        >
          <apppostthumbnail-stub
            post="[object Object]"
          />
        </nuxtlink-stub>
        <nuxtlink-stub
          class="post-link inline-block relative"
          to="/posts/10"
        >
          <apppostthumbnail-stub
            post="[object Object]"
          />
        </nuxtlink-stub>
        <nuxtlink-stub
          class="post-link inline-block relative"
          to="/posts/11"
        >
          <apppostthumbnail-stub
            post="[object Object]"
          />
        </nuxtlink-stub>
      </div>
    </section>
    <apppagination-stub
      class="justify-center my-4"
      pagination="[object Object]"
    />
  </div>
</div>
`;

exports[`pages/index renders correctly 投稿がない時 1`] = `
<div>
  <div
    post="[object Object]"
  >
    <section
      class="rounded bg-white my-8 p-4"
      data-testid="section"
    >
      <!---->
      <div
        class="text-center text-gray-700"
      >
        作品がありません
      </div>
    </section>
    <apppagination-stub
      class="justify-center my-4"
      pagination="[object Object]"
    />
  </div>
</div>
`;
